File: Eigenvalues & Eigenvectors/powerMethod.py
import numpy as np

# Function to multiply matrix and vector
def productMV(T, x):
    rows = len(T)
    cols = len(T[0])
    res = np.zeros(rows)
    for i in range(rows):
        for j in range(cols):
            res[i] += T[i][j] * x[j]
    return res

# Function to compute the norm of a vector
def norm(y):
    return np.sqrt(sum([i**2 for i in y]))

# Power Method to compute dominant eigenvalue and eigenvector
def powerMethod(T, x, r, c):
    maxIterations = 1000
    count = 0
    res = np.zeros(r)
    xi = np.array(x)
    eigen_new = 0.0
    eigen_prev = 0.0

    while True:
        # Reset eigenvalue variable
        eigen_new = 0.0
        res = productMV(T, xi)

        # Normalize the result vector
        norm1 = norm(res)
        res = res / norm1

        # Update xi to x_(i+1)
        xi = res

        # Compute eigenvalue: x^T * A * x
        intd = productMV(T, res)
        eigen_new = np.dot(intd, res)

        # Check for convergence or maximum iterations
        if abs(eigen_new - eigen_prev) < 0.001 or count > maxIterations:
            break

        eigen_prev = eigen_new
        count += 1

    print("Eigenvalue is ->", eigen_new)
    print("The computed eigenvector is:")
    for rw in res:
        print(rw)

if __name__ == "__main__":
    r = int(input("Enter the number of rows: "))
    c = int(input("Enter the number of columns: "))
    
    A = []
    print("Enter the elements row by row:")
    for i in range(r):
        row = list(map(float, input().split()))
        A.append(row)
    
    # Initial guess for eigenvector
    x = [1 if i == 0 else 0 for i in range(c)]
    # x = [1, 2, 1]
    
    powerMethod(A, x, r, c)

File: Eigenvalues & Eigenvectors/powerMethod.cpp
#include <bits/stdc++.h>
#include <cmath>
using namespace std;
#define FOR(i,a,b) for(int i=(a); i<(b); ++i)

// Note works for square matrices only..
vector<double> productMV(vector<vector<double>> T,vector<double> x){
    int rows = T.size();
    int cols = T[0].size();
    vector<double> res(rows);
    for(int i =0; i < rows; ++i){
        for(int j = 0; j < cols; ++j){
            res[i] += T[i][j]*x[j];
        }
    }
    return res;
}

double norm(vector<double> y){
    int s = y.size(), d;
    for (int i = 0; i < s; ++i){
        d += pow(y[i],2);
    }
    return sqrt(d);
}

double powerMethod(vector<vector<double>> T,vector<double> x,int r,int c){
    int maxInterations = 1000, count = 0;
    vector<double> res(r), xi = x;
     double eigen_new = 0.0,eigen_prev = 0.0;
    // for (auto e : res) cout << e << endl;

    while(true){
        //reset eigen_new variable
        eigen_new = 0.0;
        res = productMV(T,xi);

        //update x
        double norm1 = norm(res);
        for(int k = 0; k < c; ++k) {
            res[k] /= norm1;
        }

        //change xi to x_i+1
        xi = res;

        //compute eigenvalue x^t * A * x
        vector<double> intd(r);
        intd = productMV(T,res);
    
        for (int i = 0; i < r; ++i) eigen_new += intd[i]*res[i];
        // cout << "Eigenvalue is -> " << eigen_new << endl;
        
        if (abs(eigen_new-eigen_prev) < 0.001 || count > maxInterations) break;
        eigen_prev = eigen_new;
        count++;
    }
    cout << "Eigenvalue is -> " << eigen_new << endl;
    cout << "The computed eigenvector is :: " << endl;
    for (int rw = 0; rw < r; ++rw){
        cout << res[rw] << endl;
    }
}

int main(){
    int r,c;
    cout << "Enter the number of rows and cols" << endl;
    cin >> r >> c;
    vector<vector<double>> A(r,vector<double>(c));
    cout << "Enter the elements one by one" << endl;
    for (int i = 0; i < r; i++){
        for (int j = 0; j < c; j++){
            cin >> A[i][j];
            
        }
    }
    vector<double> x(c);
    //making x = {1,0,0,...}
    FOR(i,0,c) {
        if(i == 0) x[i] = 1;
        else x[i] = 0;
    }
    powerMethod(A,x,r,c);
    return 0;
}


